docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -1692,6 +1692,61 @@ _Will be added after initial implementation for PyTorch._
+
+The framework is called as Kubeflow Trainer Pipeline Framework, and it has 4 phases as you can see the following
+overview.
+
+![Overview](./TrainerPipelineFrameworkOverview.drawio.svg)
+
+As described in the following, each phase is basically executed step by step although `Startup Phase` is executed only once
+during starting trainer-controller-manager:
+
+- `Startup Phase`: Initialize internal components at once when the trainer-controller-manager starts.
+- `PreExecution Phase`: This phase is executed as a part of admission validating webhooks triggered by TrainJob is created and updated.
+- `Build Phase`: This phase is executed to build child Kubernetes resources and deploy those to the cluster.
+- `PostExecution Phase`: This phase is executed after the `Build Phase`.
+
+As you can see in the diagram, each phase has 2 types of APIs, `Internal API` and `Extension Point`.
+The Extension Point is exposed and could be added operations within the scope of the Pipeline Framework Plugins Interfaces as plugins
+and those plugins are performed in any order.
+On the other hand, the Internal APIs are not exposed and could not add any operations as opposed to the Extension Point.
+
+![Kubeflow TrainerPipelineFramework](./TrainerPipelineFramework.drawio.svg)
+
+- `Startup Phase`:
+  - Internal API:
+    - `TrainJobController`: Set up TrainJob controller and register it to Manager.
+    - `Built-in Webhook Servers`: Set up Built-in Admission Webhook Servers and register those to Manager.
+    - `Start Manager`: Start Manager.
+  - Extension Point
+    - `WatchExtension`: This registers arbitrary reconciler builders for watching any kind of resources
+    and triggering TrainJob reconciliations.
+- `PreExecution Phase`:
+  - Extension Point:
+    - `CustomValidation`: This registers validators for validating any kind of resources to Admission Validating Webhook Servers
+    when TrainJob is created and updated.
+- `Build Phase`:
+  - Internal API:
+    - `ComponentDeployer`: This deploys built components (resources) to the cluster which is performed as a part of reconciler.
+  - Extension Point:
+    - `EnforcePodGroupPolicy`: This configures PodGroup specific parameters (e.x, specified in TrainingRuntime `.spec.podGroupPolicy`)
+    to any kind of resources like PodSpec.
+    - `EnforceMLPolicy`: This configure MachineLearning framework specific parameters (e.x, specified in TrainingRuntime `.spec.mlPolicy`)
+    to any kind of resources like PodSpec.
+    - `ComponentBuilder`: This builds Kubernetes resources leveraging `RuntimeInfo` and `TrainJob`.
+    `RuntimeInfo` is abstracted objects extracted from runtimes like TrainingRuntime and ClusterTrainingRuntime.
+- `PostExecution Phase`:
+  - Internal API:
+    - `SupendedCondition`: Check if TrainJob is suspended state, and then add `Suspended` condition to TrainJob.
+    - `CreatedConditon`: Check if TrainJob is created state, and then add `Created` condition to TrainJob.
```

Comment:
Maybe more a general comment rather than something strictly in the scope of this PR, I find the name of this condition a bit confusing as the TrainJob has to be "created", even before that condition can be applied.

So either it could be stated that this condition is about whether the children components of the TrainJob have been created, or possibly rename the condition to something like `InitializedCondition`.

Reply:
> What about trying to change "Created" condition type name in a follow-up PR?

That sounds good to me 👍🏼.

Reply:
Sounds good, let's discuss it separately.
@astefanutti So do you suggest that we split Created condition for TrainJob between two ?
https://github.com/kubeflow/trainer/blob/8a6091907a2df6ab1d55b9713ead631f72ece56e/docs/proposals/2170-kubeflow-training-v2/README.md#state-transition

Reply:
@andreyvelich I was more simply thinking about renaming it to something like `Initialized` or `ComponentsCreated` more than splitting it. As it stands it may be interpreted as the TrainJob is created, which is not what it is.

Reply:
@astefanutti Please can you create an issue so we can continue discussion on how we can improve the TrainJob conditions ?
We just need to make sure that we don't confuse users given that we also have TrainJob Initializers.

Reply:
@astefanutti @andreyvelich opened: https://github.com/kubeflow/trainer/issues/2459
Let's track this enhancement in a dedicated issue.